src/components/PolicyEditor.tsx

```diff
@@ -0,0 +1,403 @@
+                                <Icon
+                                  icon="edit"
+                                  onClick={() => handleShowEditModal(el, index)}
+                                  className="color-grey mr-2"
+                                />
+                                <Icon icon="erase" onClick={() => deleteErrorHandler(index)} />
+                              </td>
+                            )}
+                          </tr>
+                        ))}
+                      </tbody>
+                    </Table>
+                  )}
+                </Form.Group>
+                <Form.Group controlId="policyRetryTimes">
+                  <Form.Label>Retry Times</Form.Label>
+                  <Form.Control
+                    type="number"
+                    autoComplete="off"
+                    required
+                    min="0"
+                    readOnly={isReadOnly}
+                    value={_.get(policyDraft, 'retry_times') || 0}
+                    onChange={handleRetryChange}
+                  />
+                </Form.Group>
+                <Form.Group controlId="policyCapacity">
+                  <Form.Label>Maximum Capacity</Form.Label>
+                  <Form.Control
+                    type="number"
+                    autoComplete="off"
+                    required
+                    min="1"
+                    readOnly={isReadOnly}
+                    value={_.get(policyDraft, 'maximum_capacity')}
+                    onChange={handlePolicyCapacityChange}
+                  />
+                </Form.Group>
+                <Form.Group controlId="policyMessageLife">
+                  <Form.Label>Message Lifetime</Form.Label>
+                  <Form.Control
+                    type="number"
+                    autoComplete="off"
+                    required
+                    min="0"
+                    readOnly={isReadOnly}
+                    value={_.get(policyDraft, 'event_ttl') || 0}
+                    onChange={handleEventTtlChange}
+                  />
+                </Form.Group>
+              </Col>
+            </Form.Row>
+          </Form>
+        </Container>
+      </Col>
+      {isSourceVisible && (
+        <Col md={6}>
+          <Form.Group controlId="policySource" className="h-100 d-flex flex-column">
+            <Form.Control
+              as="textarea"
+              className="flex-grow-1 text-monospace"
+              value={JSON.stringify(policyDraft, null, 4)}
+              onChange={handlePolicySourceChange}
+              autoComplete="off"
+              required
+              readOnly={isReadOnly}
+              isValid={isValidPolicySource}
+              isInvalid={!isValidPolicySource}
+              spellCheck={false}
+            />
+            <Form.Control.Feedback type="invalid">{policySourceError}</Form.Control.Feedback>
+          </Form.Group>
+        </Col>
+      )}
```

Comment:
The UX of the Editor can be improved. A sensible behavior could be as follows:
- Interacting with the UI fields on the left should always product valid changes to the Trigger Delivery Policy that are reported instantly on the JSON definition on the right.
- The JSON definition on the right should instead allows for free editing, so that a user can manually edit the definition without the text being automatically changed or formatted. Validation should still possibly happen on every keystroke, so the user is informed about specific errors about the definition.
- At every keystroke on the JSON definition, whenever the latter is valid, the UI fields on the left can be updated accordingly to reflect the latest valid definition that was parsed.
- As per formatting the JSON definition, we could simplify the logic as follows:
  - When the UI fields are updated, the definition is reported to the JSON on the right thanks to a call to `JSON.stringify()` in order to get a formatted text from the JavaScript's object.
  - Regarding the textual editor, we can improve it to use a custom React component instead of a simple `textarea`. It could something like [https://github.com/sujinleeme/react-json-editor](https://github.com/sujinleeme/react-json-editor), so the user has some facilities to manage a JSON text more easily, such as a dedicated *Format* or *Prettify* button to format the JSON on demand.